internal/juju/secrets.go

```diff
@@ -0,0 +1,199 @@
+}
+
+type CreateSecretInput struct {
+	ModelName string
+	Name      string
+	Value     map[string]string
+	Info      string
+}
+
+type CreateSecretOutput struct {
+	SecretId string
+}
+
+type ReadSecretInput struct {
+	SecretId  string
+	ModelName string
+	Name      *string
+	Revision  *int
+}
+
+type ReadSecretOutput struct {
+	Value map[string]string
+}
+
+type UpdateSecretInput struct {
+	SecretId  string
+	ModelName string
+	Name      *string
+	NewName   *string
+	Value     *map[string]string
+	AutoPrune *bool
+	Info      *string
+}
+
+type DeleteSecretInput struct {
+	SecretId  string
+	ModelName string
+}
+
+func newSecretsClient(sc SharedClient) *secretsClient {
+	return &secretsClient{
+		SharedClient: sc,
+		getSecretAPIClient: func(connection api.Connection) SecretAPIClient {
+			return apisecrets.NewClient(connection)
+		},
+	}
+}
+
+// CreateSecret creates a new secret.
+func (c *secretsClient) CreateSecret(input *CreateSecretInput) (CreateSecretOutput, error) {
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return CreateSecretOutput{}, err
+	}
+	defer func() { _ = conn.Close() }()
+
+	secretAPIClient := c.getSecretAPIClient(conn)
+	secretId, err := secretAPIClient.CreateSecret(input.Name, input.Info, input.Value)
+	if err != nil {
+		return CreateSecretOutput{}, typedError(err)
+	}
+	return CreateSecretOutput{
+		SecretId: secretId,
+	}, nil
+}
+
+// ReadSecret reads a secret.
+func (c *secretsClient) ReadSecret(input *ReadSecretInput) (ReadSecretOutput, error) {
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return ReadSecretOutput{}, err
+	}
+	defer func() { _ = conn.Close() }()
+
+	secretAPIClient := c.getSecretAPIClient(conn)
+
+	var secretURI *coresecrets.URI
+	if input.SecretId != "" {
+		secretURI, err = coresecrets.ParseURI(input.SecretId)
+		if err != nil {
+			return ReadSecretOutput{}, err
+		}
+	} else {
+		secretURI = nil
+	}
+	secretFilter := coresecrets.Filter{
+		URI:      secretURI,
+		Label:    input.Name,
+		Revision: input.Revision,
+	}
+
+	results, err := secretAPIClient.ListSecrets(true, secretFilter)
+	if err != nil {
+		return ReadSecretOutput{}, typedError(err)
+	}
+	if len(results) < 1 {
+		return ReadSecretOutput{}, &secretNotFoundError{secretId: input.SecretId}
+	}
+	if results[0].Error != "" {
+		return ReadSecretOutput{}, errors.New(results[0].Error)
+	}
+
+	value, err := results[0].Value.Values()
+	if err != nil {
+		return ReadSecretOutput{}, err
+	}
+	return ReadSecretOutput{
+		Value: value,
+	}, nil
+}
+
+// UpdateSecret updates a secret.
+func (c *secretsClient) UpdateSecret(input *UpdateSecretInput) error {
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	secretAPIClient := c.getSecretAPIClient(conn)
+
+	// Specify by ID or Name
+	if input.SecretId == "" && input.Name == nil {
+		return errors.New("must specify either secret ID or name")
```

Comment:
Should setting by name be allowed here? Its possible on the facade but here it just creates a mode of failure, we could just enforce that we always update by ID

Reply:
Because this is the Update method, the SecretID will always exist. Name should only be set if it's being changed by the user per the internal/provider code. Usually you'd only include UpdateSecretInput values which are being changed, not because they simply exist.

Reply:
@hmlanigan, got you. Fixed